core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -19,117 +19,57 @@
-      int fieldId = AvroSchemaUtil.getFieldId(field);
-      Types.NestedField iField = struct != null ? struct.field(fieldId) : null;
-      names.add(field.name());
-      results.add(visit(iField != null ? iField.type() : null, field.schema(), visitor));
-    }
-
-    visitor.recordLevels.pop();
-
-    return visitor.record(struct, record, names, results);
-  }
-
-  private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisitor<T> visitor) {
-    List<Schema> types = union.getTypes();
-    List<T> options = Lists.newArrayListWithExpectedSize(types.size());
-    for (Schema branch : types) {
-      if (branch.getType() == Schema.Type.NULL) {
-        options.add(visit((Type) null, branch, visitor));
-      } else {
-        options.add(visit(type, branch, visitor));
-      }
-    }
-    return visitor.union(type, union, options);
+  @Override
+  public boolean schemaEvolution() {
+    return true;
   }
 
-  private static <T> T visitArray(Type type, Schema array, AvroSchemaWithTypeVisitor<T> visitor) {
-    if (array.getLogicalType() instanceof LogicalMap || (type != null && type.isMapType())) {
-      Preconditions.checkState(
-          AvroSchemaUtil.isKeyValueSchema(array.getElementType()),
-          "Cannot visit invalid logical map type: %s", array);
-      Types.MapType map = type != null ? type.asMapType() : null;
-      List<Schema.Field> keyValueFields = array.getElementType().getFields();
-      return visitor.map(map, array,
-          visit(map != null ? map.keyType() : null, keyValueFields.get(0).schema(), visitor),
-          visit(map != null ? map.valueType() : null, keyValueFields.get(1).schema(), visitor));
-
-    } else {
-      Types.ListType list = type != null ? type.asListType() : null;
-      return visitor.array(list, array,
-          visit(list != null ? list.elementType() : null, array.getElementType(), visitor));
-    }
+  @Override
+  public boolean isMapType(Type type) {
+    return type != null && type.isMapType();
   }
 
-  private Deque<String> recordLevels = Lists.newLinkedList();
-
-  public T record(Types.StructType iStruct, Schema record, List<String> names, List<T> fields) {
-    return null;
+  @Override
+  public boolean isValidMapKey(Type type) {
+    return type == null || type instanceof Types.StringType;
   }
 
-  public T union(Type iType, Schema union, List<T> options) {
-    return null;
+  @Override
+  public Type arrayElementType(Type arrayType) {
+    return arrayType == null ? null : arrayType.asListType().elementType();
   }
 
-  public T array(Types.ListType iList, Schema array, T element) {
-    return null;
+  @Override
+  public Type mapKeyType(Type mapType) {
+    return mapType == null ? null : mapType.asMapType().keyType();
   }
 
-  public T map(Types.MapType iMap, Schema map, T key, T value) {
-    return null;
+  @Override
+  public Type mapValueType(Type mapType) {
+    return mapType == null ? null : mapType.asMapType().valueType();
```

Comment:
If this defined `isNullType`, then the visit method could do these checks instead:

```java
  if (isNullType(mapType)) {
    return nullType();
  } else {
    return mapValueType(mapType);
  }
```